cpp/src/cuda/cuda_matrix_vector_multiply.cu

```diff
@@ -0,0 +1,123 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2016 Northeastern University
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+#include "include/cuda_matrix_vector_multiply.h"
+#include <chrono>
+#define BLOCK_SIZE 16
+
+using namespace std::chrono;
+
+namespace Nice {
+
+  template <typename T>
+  __global__ void CudaMVKernel(T *d_a, T *d_x, T *d_y, int a_rows, int x_size) {
+    int row = blockIdx.y * blockDim.y + threadIdx.y;
+    int col = blockIdx.x * blockDim.x + threadIdx.x;
+    float sum = 0.0f;
+    if (row >= x_size || col >= a_rows) return;
+    for (int k = 0; k < x_size; k++) {
+      sum += (d_a[k * a_rows + col] * d_x[row * a_rows + k]);
+    }
+    __syncthreads();
+    d_y[row * a_rows + col] = sum;
+    __syncthreads();
+  }
+
+  template <typename T>
+  Vector<T> CudaMatrixVectorMultiply<T>::Multiply(const Matrix<T> &a, const Vector<T> &b) {
+    if (a.cols() == b.rows() && !a.isZero()) {
+      // Allocate and transfer memories
+      int m = a.rows();
+      int n = b.cols();
+      int k = a.cols();
+
+      const T * h_a = &a(0);
+      const T * h_x = &b(0);
+      Vector<T> h_y(m);
+
+      T * d_a;
+      T * d_x;
+      T * d_y;
+
+      // Setup GPU memory
+      CUDA_CALL(cudaMalloc(&d_a, (m * k) * sizeof(T)));
+      CUDA_CALL(cudaMemcpy(d_a, h_a, (m * k) * sizeof(T),
+        cudaMemcpyHostToDevice));
+
+      CUDA_CALL(cudaMalloc(&d_x, k * sizeof(T)));
+      CUDA_CALL(cudaMemcpy(d_x, h_x, k * sizeof(T),
+          cudaMemcpyHostToDevice));
+
+      CUDA_CALL(cudaMalloc(&d_y, m * sizeof(T)));
+      CUDA_CALL(cudaMemset(d_y, 0, m * sizeof(T)));
+
+      // Launch kernel here
+      //dim3 dimBlock(BLOCK_SIZE *BLOCK_SIZE);
+      //dim3 dimGrid((a.rows() / dimBlock.x) * (a.cols() / dimBlock.y));
+      high_resolution_clock::time_point t1 = high_resolution_clock::now();
```

Comment:
I know you are testing individual kernels here and it is hard to do it inside unit tests, what you can do is start a branch that does these timing just for your paper, while having a clean branch where timing info is removed so I can merge them into the develop branch.

Reply:
All benchmark and timing functionality will be moved to Benchmark branch.